include/unifex/when_all_range.hpp

```diff
@@ -0,0 +1,341 @@
+    type& op_;
+    void operator()() noexcept { op_.request_stop(); }
+  };
+
+  struct _operation_holder final {
+    unifex::optional<sender_nonvoid_value_type> value;
+    unifex::connect_result_t<Sender, element_receiver_t> connection;
+
+    _operation_holder(Sender&& sender, type& op, std::size_t index) noexcept(
+        unifex::is_nothrow_connectable_v<Sender, element_receiver_t>&& std::
+            is_nothrow_constructible_v<element_receiver_t, type&, std::size_t>)
+      : connection(unifex::connect(
+            static_cast<Sender&&>(sender), element_receiver_t(op, index))) {}
+  };
+
+public:
+  type(Receiver&& receiver, std::vector<Sender>&& senders)
+    : receiver_(std::move(receiver))
+    , refCount_(senders.size()) {
+    std::allocator<_operation_holder> allocator;
+    holders_ = allocator.allocate(senders.size());
+    try {
+      for (auto&& sender : senders) {
+        new (holders_ + numHolders_)
+            _operation_holder{std::move(sender), *this, numHolders_};
+        ++numHolders_;
+      }
+    } catch (...) {
+      std::destroy(holders_, holders_ + numHolders_);
+      allocator.deallocate(holders_, senders.size());
+      holders_ = nullptr;
+      throw;
+    }
+  }
+
+  type(type&&) = delete;
+
+  // does not run when constructor throws, numHolders_ is the correct size
+  ~type() {
+    std::allocator<_operation_holder> allocator;
+    std::destroy(holders_, holders_ + numHolders_);
+    allocator.deallocate(holders_, numHolders_);
+  }
+
+  void start() noexcept {
+    if (numHolders_ == 0) {
+      // In case there is 0 sender, immediately complete
+      unifex::set_value(
+          std::move(receiver_), std::vector<sender_nonvoid_value_type>{});
+    } else {
+      stopCallback_.construct(
+          unifex::get_stop_token(receiver_), cancel_operation{*this});
+      // last start() might destroy this
+      std::for_each(
+          holders_, holders_ + numHolders_, [](auto& holder) noexcept {
+            unifex::start(holder.connection);
+          });
+    }
+  }
+
+  void request_stop() noexcept {
+    if (refCount_.fetch_add(1, std::memory_order_relaxed) == 0) {
+      // deliver_result already called
+      return;
+    }
+    stopSource_.request_stop();
+
+    element_complete();
+  }
+
+  void element_complete() noexcept {
+    if (refCount_.fetch_sub(1, std::memory_order_acq_rel) == 1) {
+      stopCallback_.destruct();
+
+      if (doneOrError_.load(std::memory_order_relaxed)) {
+        if (error_.has_value()) {
+          unifex::visit(
+              [receiver = std::move(receiver_)](auto&& error) mutable {
+                unifex::set_error(
+                    std::move(receiver), std::forward<decltype(error)>(error));
+              },
+              std::move(error_.value()));
+        } else {
+          unifex::set_done(std::move(receiver_));
+        }
+      } else {
+        UNIFEX_TRY {
+          std::vector<sender_nonvoid_value_type> values;
+          std::transform(
```

Comment:
It's probably worth reserving space in `values` for the elements you're about to fill it with.
```suggestion
          values.reserve(numHolders_);
          std::transform(
```